pkg/selector/pod/selector.go

```diff
@@ -337,373 +251,88 @@ func CheckPodMeetSelector(pod v1.Pod, selector v1alpha1.PodSelectorSpec) (bool,
-		return true, nil
+func newSelectorRegistry(ctx context.Context, c client.Client, spec v1alpha1.PodSelectorSpec) registry.Registry {
+	return map[string]registry.SelectorFactory{
+		genericlabel.Name:      genericlabel.New,
+		genericnamespace.Name:  genericnamespace.New,
+		genericfield.Name:      genericfield.New,
+		genericannotation.Name: genericannotation.New,
+		nodeSelectorName: func(selector v1alpha1.GenericSelectorSpec, _ generic.Option) (generic.Selector, error) {
+			return newNodeSelector(ctx, c, spec)
+		},
+		phaseSelectorName: func(selector v1alpha1.GenericSelectorSpec, _ generic.Option) (generic.Selector, error) {
+			return newPhaseSelector(spec)
+		},
 	}
-
-	return false, nil
 }
 
-// filterByNamespaceSelector filters a list of pods by a given namespace selector.
-func filterByNamespaceSelector(pods []v1.Pod, namespaces labels.Selector) ([]v1.Pod, error) {
-	// empty filter returns original list
-	if namespaces.Empty() {
-		return pods, nil
-	}
-
-	// split requirements into including and excluding groups
-	reqs, _ := namespaces.Requirements()
-
-	var (
-		reqIncl []labels.Requirement
-		reqExcl []labels.Requirement
-
-		filteredList []v1.Pod
-	)
-
-	for _, req := range reqs {
-		switch req.Operator() {
-		case selection.Exists:
-			reqIncl = append(reqIncl, req)
-		case selection.DoesNotExist:
-			reqExcl = append(reqExcl, req)
-		default:
-			return nil, fmt.Errorf("unsupported operator: %s", req.Operator())
-		}
-	}
-
-	for _, pod := range pods {
-		// if there aren't any including requirements, we're in by default
-		included := len(reqIncl) == 0
-
-		// convert the pod's namespace to an equivalent label selector
-		selector := labels.Set{pod.Namespace: ""}
-
-		// include pod if one including requirement matches
-		for _, req := range reqIncl {
-			if req.Matches(selector) {
-				included = true
-				break
-			}
-		}
-
-		// exclude pod if it is filtered out by at least one excluding requirement
-		for _, req := range reqExcl {
-			if !req.Matches(selector) {
-				included = false
-				break
+func listPods(ctx context.Context, c client.Client, r client.Reader, spec v1alpha1.PodSelectorSpec,
+	selectorChain generic.SelectorChain, enableFilterNamespace bool) ([]v1.Pod, error) {
+	var pods []v1.Pod
+	namespaceCheck := make(map[string]bool)
+
+	if err := selectorChain.ListObjects(c, r,
+		func(listFunc generic.ListFunc, opts client.ListOptions) error {
+			var podList v1.PodList
+			if len(spec.Namespaces) > 0 {
+				for _, namespace := range spec.Namespaces {
+					if enableFilterNamespace {
+						allow, ok := namespaceCheck[namespace]
+						if !ok {
+							allow = genericnamespace.CheckNamespace(ctx, c, namespace)
+							namespaceCheck[namespace] = allow
+						}
+						if !allow {
+							continue
+						}
+					}
+
+					opts.Namespace = namespace
+					if err := listFunc(ctx, &podList, &opts); err != nil {
+						return err
+					}
+					pods = append(pods, podList.Items...)
+				}
```

Comment:
> It seems these codes have logic conflicts with `namespaceSelector` in `pkg/selector/generic/namespace/selector.go.
> 
> `namespaceSelector` assumes that items in `spec.Namespaces` should be a simple "expression" like `<namespace>`, `!<namespace>`, it supports the subset of the rules of label selector. (see pkg/selector/generic/namespace/selector.go#L103-L112)
> 
> These codes assumes that items in `spec.Namespaces` should just be the name of namespace.

It seems like a legacy issue. 😢  And I think `spec.Namespaces` should be the name of namespace here, *donotexist* operator looks a bit redundant, because the name of namespace is unique not like `label` which can has one or more objects. What do you think about it? @YangKeao

Reply:
I think it's not a problem :crying_cat_face: . I nearly forget the `namespaceSelector` supports exclude. I guess nobody uses the exclude namespace selector, because the `affectedNamespaces` in auth webhook also assumes that the `.Selector.Namespaces` are simply name of namespace, and will have problem for excluding namespace selector, but nobody has reported it.

Reply:
Wow. This function (`filterByNamespaceSelector`) exists from the very beginning of Chaos Mesh :smile_cat: https://github.com/chaos-mesh/chaos-mesh/commit/a345ea23bc0773cb7f76fc51117eb030a53bf027 .